### PR TITLE
Feat: clusteradm addon disable command for disable specified addon

### PR DESCRIPTION
**Pull request by ycyaoxdu**
Signed-off-by: ycyaoxdu yaoyuchen0626@163.com

Add a subcommand `diable` to command `addon` for enhance the new feature 3 which mentioned in  https://github.com/open-cluster-management-io/clusteradm/issues/87

### CODE REVIEW

**Comment by ycyaoxdu** [comment]
something goes wrong. https://github.com/open-cluster-management-io/clusteradm/runs/4836202667?check_suite_focus=true#step:4:65 
It seems like a bug ,need to be solved.
 i will check it out first.

**Comment by ycyaoxdu** [comment]
solved by this change https://github.com/open-cluster-management-io/clusteradm/pull/107/files#diff-57fb64e41aecc769c85303b36a57d69f497e6ba8390f3352a4f1485a029ca1d1R17 

> something goes wrong. https://github.com/open-cluster-management-io/clusteradm/runs/4836202667?check_suite_focus=true#step:4:65 It seems like a bug ,need to be solved. i will check it out first.

File: pkg/cmd/addon/disable/exec.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+		return fmt.Errorf("clusters is misisng")
+	}
+	return nil
+}
+
+func (o *Options) run() (err error) {
+	alreadyProvidedAddons := make(map[string]bool)
+	addons := make([]string, 0)
+	names := strings.Split(o.names, ",")
+	for _, n := range names {
```
Comment (qiujian16) [inline pkg/cmd/addon/disable/exec.go:43]: you can use sets in apimachinery here.

File: pkg/cmd/addon/disable/cmd.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+			if err := o.run(); err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVar(&o.names, "name", "", "Names of the add-on to deploy (comma separated)")
+	cmd.Flags().StringVarP(&o.namespace, "namespace", "n", "open-cluster-management-agent-addon", "Specified namespace to addon addon")
```
Comment (qiujian16) [inline pkg/cmd/addon/disable/cmd.go:49]: I do not think the ns and cluster here is different. You only need cluster. For `enable`, it should be the same.

File: pkg/cmd/addon/disable/exec.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+		if _, ok := alreadyProvidedAddons[n]; !ok {
+			alreadyProvidedAddons[n] = true
+			addons = append(addons, strings.TrimSpace(n))
```
Comment (qiujian16) [inline pkg/cmd/addon/disable/exec.go:46]: why append here? I think the code is easier if using sets.

File: pkg/cmd/addon/disable/cmd.go (hunk 2 above)
Comment (ycyaoxdu) [reply pkg/cmd/addon/disable/cmd.go]: ok ,will also remove that in `enable`

File: pkg/cmd/addon/disable/exec.go (hunk 1 above)
Comment (qiujian16) [reply pkg/cmd/addon/disable/exec.go:43]: you can just use addons:= sets.NewString(names...)

File: pkg/cmd/addon/disable/exec.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	names := strings.Split(o.names, ",")
+	for _, n := range names {
+		if !addons.Has(strings.TrimSpace(n)) {
+			addons.Insert(strings.TrimSpace(n))
+		}
+	}
+
+	clusters := sets.NewString()
+	cs := strings.Split(o.clusters, ",")
+	for _, c := range cs {
```
Comment (qiujian16) [inline pkg/cmd/addon/disable/exec.go:51]: clusters := sets.NewString(cs...)

File: pkg/cmd/addon/disable/cmd.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVar(&o.names, "name", "", "Names of the add-on to deploy (comma separated)")
+	cmd.Flags().StringVar(&o.clusters, "cluster", "", "Names of the managed cluster to deploy the add-on to (comma separated)")
+
```
Comment (qiujian16) [inline pkg/cmd/addon/disable/cmd.go:50]: I think we should have a`--all-clusters` command that disables addons on all clusters as a followup.

We may also do the same for enable command.

File: pkg/cmd/addon/disable/exec.go (hunk 1 above)
Comment (qiujian16) [reply pkg/cmd/addon/disable/exec.go:43]: btw, could you ensure if we did the same for enable command?

File: pkg/cmd/addon/disable/exec.go (hunk 4 above)
Comment (ycyaoxdu) [reply pkg/cmd/addon/disable/exec.go:51]: Will this change leads to repeated element, space-prefixed element or space-suffixed element to be inserted into the set if we just use `cs`, which is result of `o.clusters` split by `,`, to initialize the set. 
For example, if the input string is `--clusters cluster1,   cluster2  , cluster1 ,              cluster3`.  will this leads to an  unexpected behavior? 
I also checked the source code of set, https://github.com/kubernetes/apimachinery/blob/v0.22.1/pkg/util/sets/string.go#L30 https://github.com/kubernetes/apimachinery/blob/v0.22.1/pkg/util/sets/string.go#L49 seems that it has no duplicate element check.

Or the cobra framework already did this check for us?

Comment (qiujian16) [reply pkg/cmd/addon/disable/exec.go:51]: sets is implemented by map[key]bool so you will not have duplicated keys

Comment (ycyaoxdu) [reply pkg/cmd/addon/disable/exec.go]: > sets is implemented by map[key]bool so you will not have duplicated keys. But you certainly should trim suffix in advance

I just found that if any space exists between names, cobra will throw an error, so we don't need to trim that spaces. 
```
$ clusteradm addon enable --name test211,  test3 ,test1 --cluster cluster1
Deploying test211 add-on to namespaces open-cluster-management-agent-addon of managed cluster: cluster1.
Error: name is required
```

File: pkg/cmd/addon/disable/exec.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+)
+
+func (o *Options) complete(cmd *cobra.Command, args []string) (err error) {
+	klog.V(1).InfoS("disable options:", "dry-run", o.ClusteradmFlags.DryRun, "names", o.names, "clusters", o.clusters)
+
+	return nil
+}
+
+func (o *Options) validate() (err error) {
+	if o.names == nil {
```
Comment (qiujian16) [inline pkg/cmd/addon/disable/exec.go:29]: len(o.names) == 0

File: pkg/cmd/addon/disable/exec.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+		return fmt.Errorf("names is missing")
+	}
+
+	if o.clusters == nil {
```
Comment (qiujian16) [inline pkg/cmd/addon/disable/exec.go:33]: ditto

**Pull request merged**